core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -280,38 +280,48 @@ private[kafka] object LogValidator extends Logging {
 
     var uncompressedSizeInBytes = 0
 
-    val batch = validateOneBatchRecords(records)
+    // Assume there's only one batch with compressed memory records; otherwise, return InvalidRecordException
+    // One exception though is that with format smaller than v2, if sourceCodec is noCompression, then each batch is actually
+    // a single record so we'd need to special handle it by creating a single wrapper batch that includes all the records
+    val batches: Iterable[MutableRecordBatch] = if (sourceCodec == NoCompressionCodec && records.hasCompatibleMagic(RecordBatch.MAGIC_VALUE_V1)) {
+      records.batches.asScala
+    } else {
+      val batch = validateOneBatchRecords(records)
+      java.util.Collections.singletonList(batch).asScala
```

Comment:
nit: a bit weird to create a java collection in scala code and then covert it to scala... Seems like it's safe to just use `records.batches.asScala` since we know it contains only one entry.